Add command-line tool "librepcb-cli"

ubruhin opened the pull request:
This adds a first version of a `librepcb-cli` application which is intended to automate some frequently used tasks. It's especially useful for Continuous Integration of LibrePCB projects (e.g. with Travis-CI).

Currently it supports to load a project and...
- Report ERC messages (report failure if there are non-approved messages)
- Export schematics as PDF
- Export boards as Gerber/Excellon

For usage instructions see documentation here: https://docs.librepcb.org/_branches/add-cli/#cli

dbrgn commented on apps/librepcb-cli/commandlineinterface.cpp:
```diff
@@ -0,0 +1,334 @@
+           "there are non-approved messages)."));
+    QCommandLineOption exportSchematicsPdfOption(
+        "export-schematics-pdf",
+        tr("Export schematics as PDF to given file. Existing file will be overwritten."),
```
-> existing file**s**

ubruhin replied:
Done

dbrgn commented on apps/librepcb-cli/commandlineinterface.cpp:
```diff
@@ -0,0 +1,334 @@
+void CommandLineInterface::addToExitCode(qint16& exitCode, int count) noexcept
+{
+    exitCode = qMin(static_cast<int>(exitCode) + count, INT16_MAX);
+}
```
I'm not sure if using the exit code to convey data is a good idea. What if we would use the exit code like a bitmask? The LSB bit could indicate ERC errors, for example. Another bit could indicate files-overwritten. The error count could be printed to stderr in a well-defined format instead. That would give us more flexibility regarding the meaning of the exit code.

ubruhin replied:
Hmm yeah now I also read some advices to not use exit codes other than 0 (SUCCESS) and 1 (FAILURE). Or at least only 0-127 for our own error codes. See for example https://stackoverflow.com/questions/18248511/program-exit-status-conventions

I think I'll change it to only return 0 or 1, the error reason is printed to stderr anyway...

dbrgn replied:
I like being able to determine the type of error from the status code. But I'd try to stick within the range -128 to 127.

ubruhin replied:
What would be the use-case of that? If you want to know if the ERC failed or the Gerber export failed, just perform these tasks as two separate commands instead of one combined command:

    librepcb-cli open-project --check-erc MyProject.lpp
    librepcb-cli open-project --export-gerber MyProject.lpp

dbrgn replied:
I don't have a concrete use case, just in general I think it's nice if the exit code says something about the nature of the failure :)

dbrgn commented on apps/librepcb-cli/commandlineinterface.h:
```diff
@@ -0,0 +1,95 @@
+    public:
+
+        enum ExitCode {
+            FATAL_ERROR             = -1000,
```
TIL that the exit code is a signed integer, not a signed byte.

dbrgn commented:
In theory the "create project" dialog could ask to generate CI config files for Travis/CircleCI/Gitlab CI :slightly_smiling_face: However, that would require some kind of opt-in (since you don't want to pollute projects). Same thing would apply to the .gitignore file, which some people might not want. What do you think?

ubruhin commented:
> In theory the "create project" dialog could ask to generate CI config files for Travis/CircleCI/Gitlab CI slightly_smiling_face However, that would require some kind of opt-in (since you don't want to pollute projects). 

Yes maybe a dropdown to choose the used CI service would be nice. But I would just wait a bit with implementing this until the CLI is more mature...

In addition to changing the exit code, I think I'll also change these two things:
- Rename `--export-schematics-pdf=<file>` to `--export-schematics=<file>`. The file type is then automatically determined by the file ending. This allows to support many different formats (PDF, PNG, SVG, ...) with a single command line option (which can be given multiple times). Maybe it would some time make sense to add (optional) parameters for the export, for example something like `--export-schematics="schematics_{{PAGE}}.png,width=1024px,height=768px,background=transparent"`.
- Rename `--check-erc` to `--erc` (the "c" in "erc" stands for "check" anyway).

ubruhin commented:
Updated the branch as described above. In addition, I renamed `--export-gerber` to `--export-pcb-fabrication-data` (Gerber was not completely correct, it also generates Excellon files).

Also updated the documentation accordingly: https://docs.librepcb.org/_branches/add-cli/#cli
